### ISSUE TITLE
Support for configuring `registry_auth` directly in `docker_registry_image`

### ISSUE THREAD

**Issue by acarapetis** [kreuzwerker/terraform-provider-docker#556]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

It would be nice if the registry authentication credentials could be provided as arguments to an individual `docker_registry_image` resource, rather than set as global provider configuration. Since `docker_image` has `build.auth_config` for authentication when pulling base images, it seems like it should be possible for `docker_registry_image` to have a similar option for pushing.

### Motivation / Use Case
I've written a re-usable module that combines `aws` and `docker` to create an ECR repository, build a docker image and upload it to the ECR repository. Ideally, this module would be able to automatically acquire credentials using the `aws_ecr_authorization_token` data source (see my example below), so the module's user wouldn't need to do any configuration of the `docker` provider themselves. Since non-root Terraform modules cannot configure providers themselves (since v0.13), this is not currently possible.

### New or Affected Resource(s)

* docker_registry_image

### Potential Terraform Configuration

```hcl
data "aws_region" "current" {}
data "aws_caller_identity" "current" {}
data "aws_ecr_authorization_token" "token" {}

resource "docker_registry_image" "example" {
  name = "example"
  auth_config = {
    address  = "${data.aws_caller_identity.current.account_id}.dkr.ecr.${data.aws_region.current.name}.amazonaws.com"
    username = data.aws_ecr_authorization_token.token.user_name
    password = data.aws_ecr_authorization_token.token.password
  }
}
```

**Comment by jcogilvie** [comment]
I have a use case to:

* create a private ECR registry
* pull images from an upstream registry
* tag and push these images to the ECR registry
 
My use case would be well-served by this feature.

**Comment by Junkern** [comment]
Hey, I don't think we need to implement anything here?
I tried this locally and it works just like expected:

```
resource "docker_image" "this" {
  name = "${aws_ecr_repository.foo.repository_url}:latest"

  build {
    context = "${path.module}/src"
  }
}

resource "docker_registry_image" "registry_image" {
  name = docker_image.this.name
}

resource "aws_ecr_repository" "foo" {
  name                 = "bar"
  image_tag_mutability = "MUTABLE"

  image_scanning_configuration {
    scan_on_push = false
  }
}
data "aws_ecr_authorization_token" "token" {}
data "aws_caller_identity" "current" {}
data "aws_region" "current" {}

provider "docker" {
  registry_auth {
    address  = "${data.aws_caller_identity.current.account_id}.dkr.ecr.${data.aws_region.current.name}.amazonaws.com"
    username = data.aws_ecr_authorization_token.token.user_name
    password = data.aws_ecr_authorization_token.token.password
  }
}
```

**Comment by acarapetis** [comment]
@Junkern You're right, it does work in certain cases - terraform accepts the provider block in the context of a ["legacy shared module"](https://developer.hashicorp.com/terraform/language/modules/develop/providers#legacy-shared-modules-with-provider-configurations). Does not work if you want a modern module usable with for_each, etc.

**Comment by Junkern** [comment]
Yeah, the more I think about it the more sense it makes to also implement an auth_config for `docker_registry_image`. Thanks for the information!

**Comment by pspot2** [comment]
Was the respective data source forgotten or deliberately left out?

**Comment by Junkern** [comment]
@pspot2 I forgot it. Will be implemented with [#714](https://github.com/kreuzwerker/terraform-provider-docker/pull/714)
